proposals/distributed-async-epoch-reward.md

```diff
@@ -0,0 +1,282 @@
+computation of rewards from the actual distribution of rewards at the epoch
+boundary by dividing the process into two distinct phases:
+
+   1. rewards calculation phase - during which, the epoch reward for all
+      activate stake accounts are computed.
+
+   1. rewards distribution phase - during which, the calculated epoch rewards
+      for the active stake accounts are distributed.
+
+### Rewards Calculation
+
+Reward calculation phase computes all the rewards that need to be distributed
+for the active stake accounts, and partitions the reward into a number of
+chunks. Each of the chunks will be distributed by a separate block in the
+`reward distribution Phase`.
+
+To help track and verify the reward distribution in the next phase, two new
+sysvar accounts are introduced: (a) `EpochRewardHistory` sysvar account and (b)
+`RewardReserve` sysvar accounts. These sysvar accounts are updated at the
+boundary blocks (i.e. first and last blocks) in the reward calculation phase.
+More details of these two sysvar accounts are described in the following
+sections.
+
+The reward computation phase lasts *exactly* for `N` block heights since
+the beginning of the epoch. When reaching block height `N` after the start
+block of the `reward calculation phase`, the bank will mark it the end the
+`reward calculation phase`.
+
+Based on current validator statistics on Solana Mainnet Beta, a conservative
+value of 1,000 is chosen for `N`.
+
+At the end of the `reward calculation phase`, aka. `N` block height from the
+start of the epoch, the reward computation result is retrieved and partitioned
+into `M` chunks, which will be distributed during the `reward distribution`
+phase.
+
+Note that if the reward calculation finished before `N` block height, the
+result should be cached until `N` block height. If the reward calculation is
+slow and the result is not available yet at block-height `N`, the bank runtime
+will block and wait until the reward computation is finished, before processing
+any blocks with height greater than `N`.
+
+To ensure that each block distributes a subset of the rewards in a
+deterministic manner for the current epoch, while also randomizing the
+distribution across different epochs, the partitioning of all rewards will be
+done as follows. First, the account's public keys are hashed with the epoch
+number (i.e. sha256 hash from solana_sdk::program::hashv). Then, they are
+sorted and divided them into `M` bins. This process will ensure that the reward
+distribution is deterministic for the current epoch, while also introducing a
+degree of randomness between epochs.
+
+To minimize the impact of block time during reward distribution phase, 64
+additional accounts are stored per tick time. Given that each block time is 64
+ticks time long, a total of 4,096 accounts can be distributed per block. The
+total number of blocks needed to distributed rewards is given by the following
+formula to avoid using floating point value arithmetics.
```

Comment:
```suggestion
To minimize the impact on block processing time during the reward distribution phase, the
number of additional accounts stored per tick will be limited to 64. Given that each block is 64
ticks long, a total of 4,096 accounts can be distributed per block. The
total number of blocks needed to distributed rewards is given by the following
formula to avoid using floating point value arithmetic:
```